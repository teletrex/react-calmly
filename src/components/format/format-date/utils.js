/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */



import moment from 'moment-timezone';

import { DATE_FORMAT } from './presets';

const shortenYear = format => format.replace(/Y+/g, val => val.slice(2));

const shortenDayMonth = format => format.replace(/D+|M+/g, val => val.slice(1));

export const getFormats = (locale, longDateFormat) => {
  const format = moment.localeData(locale).longDateFormat(longDateFormat);
  const reverseFormat = format
    .split('')
    .reverse()
    .join('');
  const shortYearFormat = shortenYear(format);
  const shortDayMonthFormat = shortenDayMonth(format);
  const shortDayMonthYearFormat = shortenYear(shortDayMonthFormat);

  return {
    format,
    reverseFormat,
    shortYearFormat,
    shortDayMonthFormat,
    shortDayMonthYearFormat,
    yearOnly: 'YYYY',
  };
};

const separatorDictironary = {
  Dot: '.',
  Comma: ',',
  ForwardSlash: '/',
  Hyphen: '-',
};

const trimSeparators = format => format.replace(/^[./\-,]+|[./\-,]+$/g, '');
const replaceSeparatorBy = (format, newSeparator) => format.replace(/[.,/-]/g, newSeparator);

export const getAllSeparatorsShortDateFormats = locale => {
  const formats = getFormats(locale, DATE_FORMAT.SHORT_DATE);

  return Object.entries(separatorDictironary).reduce(
    (allFormats, [separatorName, separator]) => {
      Object.entries(formats).forEach(([formatName, format]) => {
        // eslint-disable-next-line no-param-reassign
        allFormats[`${formatName}${separatorName}`] = replaceSeparatorBy(
          trimSeparators(format),
          separator
        );
      });
      return allFormats;
    },
    { ...formats }
  );
};
