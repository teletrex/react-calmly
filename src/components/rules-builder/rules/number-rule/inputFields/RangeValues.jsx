/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React, { useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash/isEqual';
import * as Yup from 'yup';

// import NumberInput from '../../../../number-input';
import {NumberInput } from '@carbon/react';
import  {useTranslation}  from '../../../../../translation';
import RulesBuilderContext from '../../../RulesBuilderContext';
import { PROPTYPES, VALIDATION_MESSAGES } from '../../constants';
import { getPrefixedClasses, useFormValidation } from '../../helpers';

const RangeValues = ({ id, onValueChange, operator, placeholder, value }) => {
  const { t } = useTranslation();
  const { isReadOnly } = useContext(RulesBuilderContext);
  const [lastTouchedInput, setLastTouchedInput] = useState(null);

  const INPUT_ID_1 = `${id}-1`;
  const INPUT_ID_2 = `${id}-2`;

  const RANGE_VALIDATION = {
    [INPUT_ID_1]: VALIDATION_MESSAGES.rangeFirst,
    [INPUT_ID_2]: VALIDATION_MESSAGES.rangeSecond,
  };

  const validateValues = fieldValues => {
    if (fieldValues[INPUT_ID_1] >= fieldValues[INPUT_ID_2]) {
      return { [lastTouchedInput]: t(RANGE_VALIDATION[lastTouchedInput]) };
    }

    return {};
  };

  const validationSchema = Yup.object().shape({
    [INPUT_ID_1]: Yup.number(),
    [INPUT_ID_2]: Yup.number(),
  });

  const {
    errors,
    handleBlur,
    initialValues,
    resetForm,
    setFieldValue,
    submitForm,
    values,
  } = useFormValidation({
    initialValues: {
      [INPUT_ID_1]: value[0],
      [INPUT_ID_2]: value[1],
    },
    validate: validateValues,
    validationSchema,
    onSubmit: values => {
      if (!isEqual(values, initialValues)) {
        onValueChange([values[INPUT_ID_1], values[INPUT_ID_2]]);
      }
    },
  });

  useEffect(() => {
    resetForm();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [operator]);

  const handleInputBlur = e => {
    handleBlur(e);
    submitForm();
  };

  const handleValueChange = ({ imaginaryTarget: { value: newValue, id: inputId } }) => {
    setLastTouchedInput(inputId);
    setFieldValue(inputId, +newValue);
  };

  const renderReadOnly = ruleId => (
    <span className={getPrefixedClasses(['--rule-value-readonly', '--text-space-left'])}>
      {values[ruleId]}
    </span>
  );

  const renderNumberInput = (ruleId, initialValue) => (
    <NumberInput
      allowEmpty
      id={ruleId}
      invalid={!!errors[ruleId]}
      invalidText={errors[ruleId]}
      label=""
      name={ruleId}
      onBlur={handleInputBlur}
      onChange={handleValueChange}
      placeholder={placeholder || t('Enter value')}
      value={initialValue}
    />
  );

  if (isReadOnly) {
    return (
      <>
        {renderReadOnly(INPUT_ID_1)}
        <span className={getPrefixedClasses(['--text-space-left', '--text-space-right'])}>to</span>
        {renderReadOnly(INPUT_ID_2)}
      </>
    );
  }

  return (
    <>
      {renderNumberInput(INPUT_ID_1, value[0])}
      {renderNumberInput(INPUT_ID_2, value[1])}
    </>
  );
};

RangeValues.propTypes = {
  id: PropTypes.string.isRequired,
  onValueChange: PropTypes.func.isRequired,
  operator: PROPTYPES.OPERATOR,
  placeholder: PropTypes.string,
  value: PropTypes.arrayOf(PropTypes.number),
};

RangeValues.defaultProps = {
  operator: null,
  placeholder: '',
  value: [],
};

export default RangeValues;
