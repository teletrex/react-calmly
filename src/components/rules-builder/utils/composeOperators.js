/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import compact from 'lodash/compact';
import map from 'lodash/map';
import isString from 'lodash/isString';
import find from 'lodash/find';

export const composeOperators = (operators, allowedOperators) =>
  compact(
    map(operators, operator => {
      if (isString(operator)) {
        return find(allowedOperators, ({ value }) => value === operator);
      }

      return operator;
    })
  );
