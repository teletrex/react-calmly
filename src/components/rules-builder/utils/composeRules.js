/* <LICENSE>
*
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
*
* </LICENSE>
*  */




import isArray from 'lodash/isArray';
import omit from 'lodash/omit';
import uid from '@carbon/react/es/tools/uniqueId';

const mapRule = (rule, level, oneConjunctionMode) => {
  if (isArray(rule.data)) {
    let operands = 0;
    let maxLevelInside = 0;
    const data = rule.data.map(item => {
      const ret = mapRule(item, level + 1);
      operands += ret.operands;
      maxLevelInside = Math.max(maxLevelInside, ret.maxLevelInside);
      return ret;
    });
    return {
      ...rule,
      id: uid(),
      conjunction: oneConjunctionMode || rule.conjunction,
      level,
      maxLevelInside: maxLevelInside + 1,
      operands,
      data,
    };
  }

  return {
    ...rule,
    id: uid(),
    level,
    maxLevelInside: 0,
    operands: 1,
    data: rule.data,
  };
};

const mapRuleElements = (rule, createRule) => {
  const data = isArray(rule.data)
    ? rule.data.map(item => mapRuleElements(item, createRule))
    : createRule(rule.criteriaKey, rule.data, rule.timeframe);

  return {
    ...omit(rule, 'timeframe'),
    data,
  };
};

const composeRules = (rules, createRule, oneConjunctionMode) =>
  mapRuleElements(mapRule(rules, 0, oneConjunctionMode) || [], createRule);

export default composeRules;
