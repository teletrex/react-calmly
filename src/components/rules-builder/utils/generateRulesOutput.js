/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import filter from 'lodash/filter';
import isArray from 'lodash/isArray';
import isEmpty from 'lodash/isEmpty';
import map from 'lodash/map';
import omitBy from 'lodash/omitBy';
import pick from 'lodash/pick';

const generateRulesOutput = rules => {
  const { conjunction: topLevelConjunction, data: rulesData } = rules;

  const outputData = map(filter(rulesData, 'data'), rule => {
    const { data, conjunction, criteriaKey } = rule;
    const isGroup = isArray(data) && conjunction && !criteriaKey;

    return isGroup
      ? generateRulesOutput(rule)
      : omitBy(
          {
            criteriaKey,
            data: pick(data.ruleData, 'operator', 'value'),
            timeframe: pick(data.ruleTimeframe, 'operator', 'value'),
          },
          isEmpty
        );
  });

  return {
    conjunction: topLevelConjunction,
    data: outputData,
  };
};

export default generateRulesOutput;
