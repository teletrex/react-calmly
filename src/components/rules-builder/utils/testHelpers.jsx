/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';

import RulesBuilderContext from '../RulesBuilderContext';
import {
  ERROR_RELEGATION_REASON_TAG,
  relegateErrors,
} from '../../utils/test-helpers/relegate-errors';

export const datePickerErrorWrapper = fn => {
  relegateErrors(
    {
      reasonTag: ERROR_RELEGATION_REASON_TAG.FAILED_TO_RESOLVE,
      reason: "Couldn't trace the 'uncontrolled input error' in date component.",
      expectedErrorCount: 1,
    },
    () => {
      fn();
    }
  );
};

export const renderWithContext = (component, contextValue = {}) =>
  render(
    <RulesBuilderContext.Provider value={contextValue}>{component}</RulesBuilderContext.Provider>
  );
