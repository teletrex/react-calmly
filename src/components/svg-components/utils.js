/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const getSVGSizeProps = (size, height, width) => {
  const props = size ? { width: size, height: size } : {};
  if (height) {
    props.height = height;
  }
  if (width) {
    props.width = width;
  }
  return props;
};
