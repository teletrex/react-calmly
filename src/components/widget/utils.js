/* <LICENSE>
*
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
*
* </LICENSE>
*  */




import { createCssClassBuilder } from '../../utils';
import { sortObjectByValue } from '../../utils/ObjectUtils';

import { breakpoints } from './constants';

export const widgetClassBuilder = createCssClassBuilder('widget');

export const getBreakpointFromWidth = width => {
  const sortedBreakpoints = sortObjectByValue(breakpoints);

  const lowerBP = sortedBreakpoints[0][0];

  return sortedBreakpoints.reduce((currentBP, [targetBP, BPWidth]) => {
    if (width > BPWidth) {
      return targetBP;
    }
    return currentBP;
  }, lowerBP);
};

export const getGetNewPosition = layout => {
  const xByRow = { 0: 0 };
  let lowestRow = 0;

  layout.forEach(item => {
    const nextAvailableXRelativeToItem = item.x + item.w;
    const lowestItemRow = item.y + item.h - 1;

    // the item can occupy more than one row
    // save his x to each of those rows
    for (let l = 0; l < item.h; l++) {
      const currentRow = l + item.y;
      const rowX = xByRow[currentRow] || 0;
      if (nextAvailableXRelativeToItem > rowX) {
        xByRow[currentRow] = nextAvailableXRelativeToItem;
      }
    }

    if (lowestItemRow > lowestRow) {
      lowestRow = lowestItemRow;
    }
  });

  const proposedX = xByRow[lowestRow];

  return {
    // since we have 6 columns and min width is 2,
    // we cannot place an element after the 4th column,
    // we do, instead, put it on 0 becase it will land on beggining of the next line.
    x: proposedX > 4 ? 0 : proposedX,
    // always put it one line below because even if we don't need to
    // break line, react-grid-layout will vertically compress.
    y: lowestRow + 1,
  };
};
