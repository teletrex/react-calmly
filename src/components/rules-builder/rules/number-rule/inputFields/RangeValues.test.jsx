/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';

import { renderWithContext } from '../../../utils/testHelpers';

import RangeValues from './RangeValues';

const initialValues = [3, 7];
const expectedValues = initialValues.map(String);

const props = {
  id: 'TestNumberRuleElement',
  onValueChange: jest.fn(),
};

describe('RangeValues', () => {
  beforeEach(() => jest.resetAllMocks());

  it('should render correctly', () => {
    const { container } = render(<RangeValues {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with a provided value', () => {
    const { container, getByDisplayValue } = render(
      <RangeValues {...props} value={initialValues} />
    );
    expect(container).toMatchSnapshot();
    expect(getByDisplayValue(expectedValues[0])).toBeInTheDocument();
    expect(getByDisplayValue(expectedValues[1])).toBeInTheDocument();
  });

  it('should render correctly for read-only', () => {
    const { container, getByText } = renderWithContext(
      <RangeValues {...props} value={initialValues} />,
      { isReadOnly: true }
    );
    expect(container).toMatchSnapshot();
    expect(getByText(expectedValues[0])).toBeInTheDocument();
    expect(getByText(expectedValues[1])).toBeInTheDocument();
  });

  it('should properly set the rule value when range values are entered and input fields are blurred', async () => {
    const { getAllByPlaceholderText } = render(<RangeValues {...props} />);
    const numberInputs = getAllByPlaceholderText('Enter value');

    fireEvent.change(numberInputs[0], { target: { value: initialValues[0] } });
    await waitFor(() => expect(props.onValueChange).toHaveBeenCalledTimes(0));
    fireEvent.blur(numberInputs[0]);
    await waitFor(() =>
      expect(props.onValueChange).toHaveBeenCalledWith([initialValues[0], undefined])
    );

    props.onValueChange.mockReset();

    fireEvent.change(numberInputs[1], { target: { value: initialValues[1] } });
    await waitFor(() => expect(props.onValueChange).toHaveBeenCalledTimes(0));
    fireEvent.blur(numberInputs[1]);
    await waitFor(() => expect(props.onValueChange).toHaveBeenCalledWith(initialValues));
  });

  it('should display an error when the first value entered is greater than the second', async () => {
    const { findByText, getAllByPlaceholderText } = render(
      <RangeValues {...props} value={initialValues} />
    );
    const numberInputs = getAllByPlaceholderText('Enter value');

    fireEvent.change(numberInputs[0], { target: { value: initialValues[1] + 1 } });
    fireEvent.blur(numberInputs[0]);

    const errorMessage = await findByText('The first value must be smaller than the second.');

    expect(errorMessage).toBeInTheDocument();
  });

  it('should display an error when the second value entered is smaller than the first', async () => {
    const { findByText, getAllByPlaceholderText } = render(
      <RangeValues {...props} value={initialValues} />
    );
    const numberInputs = getAllByPlaceholderText('Enter value');

    fireEvent.change(numberInputs[1], { target: { value: initialValues[0] - 1 } });
    fireEvent.blur(numberInputs[1]);

    const errorMessage = await findByText('The second value must be greater than the first.');

    expect(errorMessage).toBeInTheDocument();
  });

  it('should not update the rule if the values were not changed', async () => {
    const { getAllByPlaceholderText } = render(<RangeValues {...props} value={initialValues} />);
    const numberInputs = getAllByPlaceholderText('Enter value');

    fireEvent.blur(numberInputs[0]);
    fireEvent.blur(numberInputs[1]);

    await waitFor(() => expect(props.onValueChange).toHaveBeenCalledTimes(0));
  });
});
