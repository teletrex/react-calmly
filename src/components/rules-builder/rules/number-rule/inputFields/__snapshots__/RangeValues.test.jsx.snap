// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeValues should render correctly 1`] = `
<div>
  <div
    class="cds--number-input-wrapper"
  >
    <div
      class="cds--form-item"
    >
      <div
        class="cds--number cds--number--helpertext cds--number-input-wrapper_carbon-input"
      >
        <div
          class="cds--number__input-wrapper"
        >
          <input
            aria-label="Numeric input field with increment and decrement buttons"
            id="TestNumberRuleElement-1"
            max="Infinity"
            min="-Infinity"
            name="TestNumberRuleElement-1"
            pattern="[0-9]*"
            placeholder="Enter value"
            step="1"
            type="number"
            value=""
          />
          <div
            class="cds--number__controls"
          >
            <button
              aria-atomic="true"
              aria-label="Increment number. Press Space or Enter in order to use icon."
              aria-live="polite"
              class="cds--number__control-btn up-icon"
              focusable="false"
              tabindex="-1"
              title="Increment number. Press Space or Enter in order to use icon."
              type="button"
              unselectable="on"
            >
              <svg
                aria-hidden="true"
                class="up-icon"
                focusable="false"
                height="4"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 8 4"
                width="8"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 4l4-4 4 4z"
                />
              </svg>
            </button>
            <button
              aria-atomic="true"
              aria-label="Decrement number. Press Space or Enter in order to use icon."
              aria-live="polite"
              class="cds--number__control-btn down-icon"
              focusable="false"
              tabindex="-1"
              title="Decrement number. Press Space or Enter in order to use icon."
              type="button"
              unselectable="on"
            >
              <svg
                aria-hidden="true"
                class="down-icon"
                focusable="false"
                height="4"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 8 4"
                width="8"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 0L4 4 0 0z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>

  </div>
  <div
    class="cds--number-input-wrapper"
  >
    <div
      class="cds--form-item"
    >
      <div
        class="cds--number cds--number--helpertext cds--number-input-wrapper_carbon-input"
      >
        <div
          class="cds--number__input-wrapper"
        >
          <input
            aria-label="Numeric input field with increment and decrement buttons"
            id="TestNumberRuleElement-2"
            max="Infinity"
            min="-Infinity"
            name="TestNumberRuleElement-2"
            pattern="[0-9]*"
            placeholder="Enter value"
            step="1"
            type="number"
            value=""
          />
          <div
            class="cds--number__controls"
          >
            <button
              aria-atomic="true"
              aria-label="Increment number. Press Space or Enter in order to use icon."
              aria-live="polite"
              class="cds--number__control-btn up-icon"
              focusable="false"
              tabindex="-1"
              title="Increment number. Press Space or Enter in order to use icon."
              type="button"
              unselectable="on"
            >
              <svg
                aria-hidden="true"
                class="up-icon"
                focusable="false"
                height="4"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 8 4"
                width="8"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 4l4-4 4 4z"
                />
              </svg>
            </button>
            <button
              aria-atomic="true"
              aria-label="Decrement number. Press Space or Enter in order to use icon."
              aria-live="polite"
              class="cds--number__control-btn down-icon"
              focusable="false"
              tabindex="-1"
              title="Decrement number. Press Space or Enter in order to use icon."
              type="button"
              unselectable="on"
            >
              <svg
                aria-hidden="true"
                class="down-icon"
                focusable="false"
                height="4"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 8 4"
                width="8"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 0L4 4 0 0z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>

  </div>
</div>
`;

exports[`RangeValues should render correctly for read-only 1`] = `
<div>
  <span
    class="cds--rule-value-readonly cds--text-space-left"
  >
    3
  </span>
  <span
    class="cds--text-space-left cds--text-space-right"
  >
    to
  </span>
  <span
    class="cds--rule-value-readonly cds--text-space-left"
  >
    7
  </span>
</div>
`;

exports[`RangeValues should render correctly with a provided value 1`] = `
<div>
  <div
    class="cds--number-input-wrapper"
  >
    <div
      class="cds--form-item"
    >
      <div
        class="cds--number cds--number--helpertext cds--number-input-wrapper_carbon-input"
      >
        <div
          class="cds--number__input-wrapper"
        >
          <input
            aria-label="Numeric input field with increment and decrement buttons"
            id="TestNumberRuleElement-1"
            max="Infinity"
            min="-Infinity"
            name="TestNumberRuleElement-1"
            pattern="[0-9]*"
            placeholder="Enter value"
            step="1"
            type="number"
            value="3"
          />
          <div
            class="cds--number__controls"
          >
            <button
              aria-atomic="true"
              aria-label="Increment number. Press Space or Enter in order to use icon."
              aria-live="polite"
              class="cds--number__control-btn up-icon"
              focusable="false"
              tabindex="-1"
              title="Increment number. Press Space or Enter in order to use icon."
              type="button"
              unselectable="on"
            >
              <svg
                aria-hidden="true"
                class="up-icon"
                focusable="false"
                height="4"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 8 4"
                width="8"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 4l4-4 4 4z"
                />
              </svg>
            </button>
            <button
              aria-atomic="true"
              aria-label="Decrement number. Press Space or Enter in order to use icon."
              aria-live="polite"
              class="cds--number__control-btn down-icon"
              focusable="false"
              tabindex="-1"
              title="Decrement number. Press Space or Enter in order to use icon."
              type="button"
              unselectable="on"
            >
              <svg
                aria-hidden="true"
                class="down-icon"
                focusable="false"
                height="4"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 8 4"
                width="8"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 0L4 4 0 0z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>

  </div>
  <div
    class="cds--number-input-wrapper"
  >
    <div
      class="cds--form-item"
    >
      <div
        class="cds--number cds--number--helpertext cds--number-input-wrapper_carbon-input"
      >
        <div
          class="cds--number__input-wrapper"
        >
          <input
            aria-label="Numeric input field with increment and decrement buttons"
            id="TestNumberRuleElement-2"
            max="Infinity"
            min="-Infinity"
            name="TestNumberRuleElement-2"
            pattern="[0-9]*"
            placeholder="Enter value"
            step="1"
            type="number"
            value="7"
          />
          <div
            class="cds--number__controls"
          >
            <button
              aria-atomic="true"
              aria-label="Increment number. Press Space or Enter in order to use icon."
              aria-live="polite"
              class="cds--number__control-btn up-icon"
              focusable="false"
              tabindex="-1"
              title="Increment number. Press Space or Enter in order to use icon."
              type="button"
              unselectable="on"
            >
              <svg
                aria-hidden="true"
                class="up-icon"
                focusable="false"
                height="4"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 8 4"
                width="8"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 4l4-4 4 4z"
                />
              </svg>
            </button>
            <button
              aria-atomic="true"
              aria-label="Decrement number. Press Space or Enter in order to use icon."
              aria-live="polite"
              class="cds--number__control-btn down-icon"
              focusable="false"
              tabindex="-1"
              title="Decrement number. Press Space or Enter in order to use icon."
              type="button"
              unselectable="on"
            >
              <svg
                aria-hidden="true"
                class="down-icon"
                focusable="false"
                height="4"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 8 4"
                width="8"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 0L4 4 0 0z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>

  </div>
</div>
`;
